Storybook: add custom tooltip

CoffeeGeek101 opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description

Adding Tooltip implementation to the storybook, Simple tooltip, Custom single valued tooltip and Custom multiple valued tooltip

<!--- Describe your changes in detail -->

A Tooltip story has been added to the API section of the Storybook

**Before :**
<img  width="227" alt="Screenshot 2023-04-30 at 11 36 15 PM" src="https://user-images.githubusercontent.com/110016849/235369163-92871962-4be4-4e4c-b34c-b96f7987464b.png">
**After :**
<img  width="229" alt="Screenshot 2023-04-30 at 11 36 29 PM" src="https://user-images.githubusercontent.com/110016849/235369199-a8352af5-429d-4e2d-b292-6228dd3c0afb.png">

<!--- This project only accepts pull requests related to open issues -->
<!--- If suggesting a new feature or change, please discuss it in an issue first -->
<!--- If fixing a bug, there should be an issue describing it with steps to reproduce -->
<!--- Please link to the issue here: -->

## Motivation and Context

<!--- Why is this change required? What problem does it solve? -->
The default tooltip of the library is good enough, but user might want to have the tooltip matches with their UI, and there is no direct prop available to customised the component, might lead to a bad UX.

## How Has This Been Tested?

<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
This is an initial PR just to get the work started toward it, so the changes haven't tested yet.
## Screenshots (if appropriate):

This is directory structure and the custom tooltip implementation preview:

<img width="605" alt="Screenshot 2023-04-30 at 11 44 07 PM" src="https://user-images.githubusercontent.com/110016849/235369472-9b8e9129-bedc-450b-ad8f-133aa805acbb.png">


## Types of changes

<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

CoffeeGeek101 commented:
Need a feedback on the work till now, If the implementation is alright or not.

ckifer commented on storybook/stories/API/tooltip/Tooltip.stories.tsx:
```diff
@@ -0,0 +1,71 @@
+            color: '#fff',
+          }}
+        >
+          {`${label} : ${payload[0].value}`}
```
would do some validation that `${payload[0].value}` exists or use `payload?.[0]?.value}`

ckifer commented on storybook/stories/API/tooltip/Tooltip.stories.tsx:
```diff
@@ -0,0 +1,71 @@
+  },
+};
+
+const CustomTooltip = ({ active, payload, label }: any) => {
```
Can you type this? Should return `JSX.Element | null` I think (or similar)?

ckifer commented on storybook/stories/API/tooltip/Tooltip.stories.tsx:
```diff
@@ -0,0 +1,71 @@
+  if (active && payload) {
+    return (
+      <div
+        // className="custom-tooltip"
```
either add these or remove them. I think they are fine to be there as classNames for an example

ckifer reviewed (commented):
Thanks for the contribution, looking good so far

CoffeeGeek101 commented:
Heyy, @ckifer thank you feedback, I have fixed the type and added the validation to the payload array.
Also, removed the 'className' comments, which are not required here.

If this is alright, I'll start working on the multiple valued tooltip story and docs mdx file for them.

nikolasrieble commented on storybook/stories/API/tooltip/Tooltip.stories.tsx:
```diff
@@ -0,0 +1,69 @@
+  component: LineChart,
+};
+
+export const SimpleTooltip = {
+  render: (args: Record<string, any>) => {
+    return (
+      <ResponsiveContainer width="100%" height={400}>
+        <LineChart {...args}>
+          <Line dataKey="uv" />
+          <Tooltip />
+        </LineChart>
+      </ResponsiveContainer>
+    );
+  },
+  args: {
+    data: pageData,
+    dataKey: 'uv',
+  },
+};
```
```suggestion
export const SimpleTooltip = {
  render: (tooltipArgs: Record<string, any>) => {
    return (
      <ResponsiveContainer width="100%" height={400}>
        <LineChart data={pageData}>
          <Line dataKey="uv" />
          <Tooltip {...tooltipArgs} />
        </LineChart>
      </ResponsiveContainer>
    );
  },
};
```

In the API section of storybook, we want to document the API of a single component, in this case, the Tooltip. 

Thus every story should only have args that belong to the component in question.

nikolasrieble replied:
You could consider the Line stories as an example. There, we group the props by categories (i.e. animation, style, general, ...) and create a story for each, with non default values, which show the behaviour and are changeable per story.

CoffeeGeek101 replied:
okay, let me explain the suggested changes, so that I can be certain if I understood the required changes to be made.

- the tooltip story currently having args that is not relevant to the component, so I need to change the args that are specific to the tooltip component and will displayed in the controls section of the storybook playground.

- If I understand the problem correctly, I need to use "tooltipArgs" for that case.

nikolasrieble replied:
Yes, makes sense.

CoffeeGeek101 replied:
I was checking out the Line stories, and understood the problem in this PR, but I couldn't find any tooltipArgs on the API props, do I need to or can I import **TooltipProps** from the tooltip component

nikolasrieble replied:
tooltipPops is not a type just an alias to make clear that we only want to pass it to the tooltip in api stories

nikolasrieble reviewed (commented):
Great first step! Please consider my suggestion

nikolasrieble commented on storybook/stories/API/tooltip/Tooltip.stories.tsx:
```diff
@@ -0,0 +1,69 @@
+  render: (tootipArgs: Record<string, any>) => {
+    return (
+      <ResponsiveContainer width="100%" height={400}>
+        <LineChart {...tootipArgs}>
+          <Line dataKey="uv" />
+          <Tooltip />
+        </LineChart>
```
```suggestion
        <LineChart data={pageData} >
          <Line dataKey="uv" />
          <Tooltip  {...tootipArgs} />
        </LineChart>
```
The args must still be passed to the Tooltip component.

nikolasrieble replied:
And the chart needs data.

CoffeeGeek101 replied:
Thank you, I really made a massive mistake there. Working on it!

nikolasrieble commented on storybook/stories/API/tooltip/Tooltip.stories.tsx:
```diff
@@ -0,0 +1,69 @@
+      </ResponsiveContainer>
+    );
+  },
+  args: {
+    data: pageData,
+    dataKey: 'uv',
+  },
```
```suggestion
```

Here no args are needed.

nikolasrieble commented on storybook/stories/API/tooltip/Tooltip.stories.tsx:
```diff
@@ -0,0 +1,69 @@
+};
+
+export const CustomSingleValued = {
+  render: (tooltipArgs: Record<string, any>) => {
+    return (
+      <ResponsiveContainer width="100%" height={400}>
+        <LineChart {...tooltipArgs}>
+          <Line dataKey="uv" />
+          <Tooltip content={<CustomTooltip />} />
+        </LineChart>
+      </ResponsiveContainer>
+    );
+  },
+  args: {
+    data: pageData,
+    dataKey: 'uv',
+  },
```
If you change the first story following my suggestion, you might then simply reuse the render function as such: 

```suggestion
  ... SimpleTooltip,
  args: {
    content: <CustomTooltip />,
  },
```

nikolasrieble replied:
Because between the first and the second story, the only difference is the value passed to the content prop.

CoffeeGeek101 replied:
This really helped me, I understood, what I was doing wrong. I have made the possible changes.

CoffeeGeek101 commented on storybook/stories/API/tooltip/Tooltip.stories.tsx:
```diff
@@ -0,0 +1,69 @@
+  },
+};
+
+const CustomTooltip: React.FC = ({ active, payload, label }: any) => {
```
I have added some types for the props

<img width="596" alt="Screenshot 2023-05-06 at 11 07 17 AM" src="https://user-images.githubusercontent.com/110016849/236602486-f6faebc9-4c62-4c41-a16c-38786ce7d44e.png">

CoffeeGeek101 commented on storybook/stories/API/tooltip/Tooltip.stories.tsx:
```diff
@@ -0,0 +1,69 @@
+            color: '#fff',
+          }}
+        >
+          {`${label} : ${payload[0]?.value}`}
```
I have handle the case when the payload array is empty and updated the label with the key of value that is coming from the payload array.

<img width="612" alt="Screenshot 2023-05-06 at 11 08 31 AM" src="https://user-images.githubusercontent.com/110016849/236602587-e07cff20-f257-4f63-ad63-ae8a274c3d16.png">

CoffeeGeek101 commented:
If the single value is tooltip is ready, can I start with the multi - valued tooltip?

CoffeeGeek101 commented:
I'm getting this eslint error, can you give some context why this might be happening?
<img width="812" alt="Screenshot 2023-05-11 at 1 28 01 PM" src="https://github.com/recharts/recharts/assets/110016849/12276218-9afa-4b88-95c6-9def2bd843bf">

kolserdav commented:
Is there an example of use this feature?

ckifer commented:
@kolserdav  not sure what you mean. This is a PR that added some changes to the storybook rather than a feature. Those examples are deployed on the storybook if you want to review them there. Link in the readme
